src/genkey/clu_genkey.c

```diff
@@ -1274,6 +1274,247 @@ int wolfCLU_genKey_Dilithium(WC_RNG* rng, char* fName, int directive, int fmt,
+    char* fOutNameBuf = NULL;
+
+#ifdef NO_AES
+    /* use 16 bytes for AES block size */
+    size_t maxDerBufSz = 4 * keySz * 16;
+#else
+    size_t maxDerBufSz = 4 * keySz * AES_BLOCK_SIZE;
+#endif  /* NO_AES */
+
+    byte*  derBuf      = NULL;
+    byte*  pemBuf      = NULL;
+    byte*  outBuf      = NULL;
+    int    derBufSz    = -1;
+    int    pemBufSz    = 0;
+    int    outBufSz    = 0;
+
+#ifdef WOLFSSL_SMALL_STACK
+    MlDsaKey* key;
+    key = (MlDsaKey*)XMALLOC(sizeof(MlDsaKey), HEAP_HINT,
+            DYNAMIC_TYPE_DILITHIUM);
+    if (key == NULL) {
+        return MEMORY_E;
+    }
+#else
+    MlDsaKey key[1];
+#endif
+
+    if (rng == NULL || fName == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* init the ML-DSA key */
+    if (wc_MlDsaKey_Init(key, NULL, 0) != 0) {
+        wolfCLU_LogError("Failed to initialize ML-DSA Key.\nRET: %d", ret);
+#ifdef WOLFSSL_SMALL_STACK
+        XFREE(key, HEAP_HINT, DYNAMIC_TYPE_DILITHIUM);
+#endif
+        return ret;
+    }
+    XMEMSET(key, 0, sizeof(MlDsaKey));
+
+    /* set the level of the ML-DSA key */
+    if (wc_MlDsaKey_SetParams(key, level) != 0) {
+        wc_MlDsaKey_Free(key);
+#ifdef WOLFSSL_SMALL_STACK
+        XFREE(key, HEAP_HINT, DYNAMIC_TYPE_DILITHIUM);
+#endif
+        return WOLFCLU_FAILURE;
+    }
+
+    /* make the ML-DSA key */
+    if (wc_MlDsaKey_MakeKey(key, rng) != 0) {
+        wc_MlDsaKey_Free(key);
+#ifdef WOLFSSL_SMALL_STACK
+        XFREE(key, HEAP_HINT, DYNAMIC_TYPE_DILITHIUM);
+#endif
+        return WOLFCLU_FAILURE;
+    }
+
+    /* set up the file name output buffer */
+    if (ret == WOLFCLU_SUCCESS) {
+        fNameSz     = (int)XSTRLEN(fName);
+        fOutNameBuf = (char*)XMALLOC(fNameSz + fExtSz + 1, HEAP_HINT,
+                        DYNAMIC_TYPE_TMP_BUFFER);
+        if (fOutNameBuf == NULL) {
+            ret = MEMORY_E;
+        }
+    }
+
+    if (ret == WOLFCLU_SUCCESS) {
+        XMEMSET(fOutNameBuf, 0, fNameSz + fExtSz);
+        XMEMCPY(fOutNameBuf, fName, fNameSz);
+
+        derBuf = (byte*)XMALLOC(maxDerBufSz, HEAP_HINT,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+        if (derBuf == NULL) {
+            ret = MEMORY_E;
+        }
+    }
+
+    if (ret == WOLFCLU_SUCCESS) {
+        switch (directive) {
+            case PRIV_AND_PUB_FILES:
+                /* Fall through to PRIV_ONLY_FILE */
+                FALL_THROUGH;
+            case PRIV_ONLY_FILE:
+                /* add on the final part of the file name ".priv" */
+                XMEMCPY(fOutNameBuf + fNameSz, fExtPriv, fExtSz);
+                WOLFCLU_LOG(WOLFCLU_L0, "Private key file = %s", fOutNameBuf);
+
+                /* Private key to der */
+                derBufSz = wc_MlDsaKey_PrivateKeyToDer(key,
+                                derBuf, (word32)maxDerBufSz);
+                if (derBufSz < 0) {
+                    ret = derBufSz;
+                }
+                outBuf   = derBuf;
+                outBufSz = derBufSz;
```

Comment:
Here the `outBufSz` could end up being negative if the value `derBufSz` is negative. Likely that's not causing trouble for the caller of this function, but I'd prefer that outBufSz be 0 on failure to indicate there is not output buffer data since it failed.

Reply:
Similar case with pemBufSz in the next code section.